DIPs/1NNN-ADR.md

```diff
@@ -0,0 +1,479 @@
+
+In Phobos
+```
+auto writefln(Fmt, Args...)(Fmt fmt, Args args) if(is(Fmt == _d_interpolated_string!Ignored, Ignored...)) {
+        import std.stdio; // of course this should be IN std.stdio, making this import unnecessary
+        return std.stdio.writefln!(fmt.toFormatString!"%s", Args)(args);
+}
+// ditto for formattedWrite, format, and any other uses.
+```
+
+These ensure that interpolated strings just work for their most likely target functions while also providing a new benefit: the format string, including user additions via `${}`, will be checked at compile time, even when passed as a runtime argument.
+
+```
+string name;
+writefln(i"Hello, ${%d}name"); // causes a compile error for invalid format specification
+```
+
+Note that this comes despite the compiler itself knowing absolutely nothing about format specifiers - it is entirely implemented in library code, and third parties may provide their own formats and checks.
+
+We may also provide user-friendly aliases for `_d_interpolated_string` and `_d_interpolated_format_spec`, but the color of the bikeshed is not important to this DIP author.
+
+### Justifications
+
+The library wrapper may seem superfluous, however it serves three key roles:
+
+1. It divorces the compiler entirely from details of generated format strings.
+2. It provides necessary error checking capabilities.
+3. It provides an additional API for user functions to introspect the string, building on D's existing compile-time capabilities.
+
+The previous version of this DIP proposed a simple string literal instead of the result of `_d_interpolated_string` as the first argument passed. This is problematic because it introduces potentially subtle errors in likely usage.
+
+#### Wrong-use in unrelated function
+
+Consider a function in D today:
+
+```
+/++
+	Creates a window with the given title, width, and height. If
+	width or height are set to zero (the default), your window will
+	be automatically sized.
++/
+Window createWindow(string title, int width = 0, int height = 0);
+```
+
+A user has some code that calls it like so:
+
+```
+import std.conv;
+auto window = createWindow("Process debugger " ~ to!string(pid));
+```
+
+A new version of the compiler is released and the user, eager to try D's new string interpolation feature, rewrites that code as would be common in many other languages:
+
+```
+// look at this clean new code :D
+auto window = createWindow(i"Process debugger $pid");
+```
+
+It compiles without error. In bliss, the happy user thinks: "d rox". Then... the program is run and the window width is extraordinary. But the documentation says it will be automatically sized by default. The confused user wonders: why did this change? They run the program again and get another random width. Is the library incompatible with the new dmd version?
```

Comment:
Be more technical and less anecdotal here. What unintended errors would have occurred if this hypothetical string interpolation feature were to be implemented? How does it compare to you implementation?